--- a/src/video/SDL_egl.c
+++ b/src/video/SDL_egl.c
@@ -19,6 +19,7 @@
  *  3. This notice may not be removed or altered from any source distribution.
  */
 #include "../SDL_internal.h"
+#include <unistd.h>
 
 #if SDL_VIDEO_OPENGL_EGL
 
--- a/src/video/raspberry/SDL_rpivideo.h
+++ b/src/video/raspberry/SDL_rpivideo.h
@@ -30,6 +30,12 @@
 #include "EGL/egl.h"
 #include "EGL/eglext.h"
 
+typedef struct {
+   DISPMANX_ELEMENT_HANDLE_T element;
+   int width;   /* This is necessary because dispmanx elements are not queriable. */
+   int height;
+} EGL_DISPMANX_WINDOW_T;
+
 typedef struct SDL_VideoData
 {
     uint32_t egl_refcount;      /* OpenGL ES reference count              */

