diff --git a/wiringPi/wiringPi.c b/wiringPi/wiringPi.c
index bd1a369..a25743e 100644
--- a/wiringPi/wiringPi.c
+++ b/wiringPi/wiringPi.c
@@ -690,7 +690,7 @@ int piGpioLayout (void)
   FILE *cpuFd ;
   char line [120] ;
   char *c ;
-  static int  gpioLayout = -1 ;
+  static int  gpioLayout = 2 ; // No Hardware line in /proc/cpuinfo for RPi3 in 64bit mode
 
   if (gpioLayout != -1)	// No point checking twice
     return gpioLayout ;
