--- a/make.conf	2018-09-21 12:55:28.250264229 +0100
+++ b/make.conf	2018-09-21 13:00:20.784160827 +0100
@@ -9,12 +9,12 @@
 # Some scripts looks for PORTDIR in make.conf explicitly
 PORTDIR="/usr/portage"
 
-# for use when compiling locally
-MAKEOPTS="-j5 -l4"
-EMERGE_DEFAULT_OPTS="--jobs=5 --load-average=4"
-# for use with compiling with distcc only
-#MAKEOPTS="-j8 -l4"
-#EMERGE_DEFAULT_OPTS="--jobs=5 --load-average=4"
+# override default build parallelism by uncommenting the following
+# two lines, and modifying the values as required
+#MAKEOPTS="-j2 -l1"
+#EMERGE_DEFAULT_OPTS="--jobs=2 --load-average=1"
+# caution! many parallel threads and / or emerge jobs may cause your
+# RPi3 to run out of memory during large builds, or genup runs
 
 # per https://wiki.gentoo.org/wiki/Raspberry_Pi_VC4
 VIDEO_CARDS="fbdev vc4"
