diff --git a/media/learning/common/value.h b/media/learning/common/value.h
index 62f4953..ef37eeb 100644
--- a/media/learning/common/value.h
+++ b/media/learning/common/value.h

@@ -27,7 +27,7 @@
  public:
   Value();
   template <typename T>
-  explicit Value(T x) : value_(x) {
+  explicit Value(const T& x) : value_(x) {
     // We want to rule out mostly pointers, since they wouldn't make much sense.
     // Note that the implicit cast would likely fail anyway.
     static_assert(std::is_arithmetic<T>::value || std::is_enum<T>::value,
