--- a/third_party/skia/src/opts/SkRasterPipeline_opts.h.aarch64 2018-08-21 11:51:23.000000000 +0200
+++ b/third_party/skia/src/opts/SkRasterPipeline_opts.h 2018-08-27 01:46:51.731550345 +0200
@@ -657,10 +657,7 @@
 }
 
 SI F from_half(U16 h) {
-#if defined(__aarch64__) && !defined(SK_BUILD_FOR_GOOGLE3)  // Temporary workaround for some Google3 builds.
-    return vcvt_f32_f16(h);
-
-#elif defined(JUMPER_IS_HSW) || defined(JUMPER_IS_AVX512)
+#if defined(JUMPER_IS_HSW) || defined(JUMPER_IS_AVX512)
     return _mm256_cvtph_ps(h);
 
 #else
@@ -677,10 +674,7 @@
 }
 
 SI U16 to_half(F f) {
-#if defined(__aarch64__) && !defined(SK_BUILD_FOR_GOOGLE3)  // Temporary workaround for some Google3 builds.
-    return vcvt_f16_f32(f);
-
-#elif defined(JUMPER_IS_HSW) || defined(JUMPER_IS_AVX512)
+#if defined(JUMPER_IS_HSW) || defined(JUMPER_IS_AVX512)
     return _mm256_cvtps_ph(f, _MM_FROUND_CUR_DIRECTION);
 
 #else
--- chromium-70.0.3528.4/third_party/crc32c/src/src/./crc32c_arm64_linux_check.h.aarch64        2018-08-21 11:51:17.000000000 +0200
+++ chromium-70.0.3528.4/third_party/crc32c/src/src/./crc32c_arm64_linux_check.h        2018-08-27 03:33:34.550997403 +0200
@@ -33,7 +33,15 @@
   // From 'arch/arm64/include/uapi/asm/hwcap.h' in Linux kernel source code.
   constexpr unsigned long kHWCAP_PMULL = 1 << 4;
   constexpr unsigned long kHWCAP_CRC32 = 1 << 7;
-  unsigned long hwcap = (&getauxval != nullptr) ? getauxval(AT_HWCAP) : 0;
+  unsigned long hwcap =
+#if !HAVE_STRONG_GETAUXVAL
+    (&getauxval != nullptr) ?
+#endif
+    getauxval(AT_HWCAP)
+#if !HAVE_STRONG_GETAUXVAL
+    : 0
+#endif
+    ;
   return (hwcap & (kHWCAP_PMULL | kHWCAP_CRC32)) ==
          (kHWCAP_PMULL | kHWCAP_CRC32);
 #else

