--- a/sh/init.sh.Linux.in
+++ b/sh/init.sh.Linux.in
@@ -51,6 +51,11 @@
 	eend $?
 fi
 
+# fix annoying clock skew messages, if swclock service in use
+# see https://gitlab.alpinelinux.org/alpine/aports/issues/8093
+[ -h /etc/runlevels/boot/swclock ] && \
+	"$RC_LIBEXECDIR/sbin/swclock" "/etc/init.d"
+
 # /run is a new directory for storing volatile runtime data.
 # Read more about /run at https://lwn.net/Articles/436012
 sys="$(openrc --sys)"
