diff -uNr a/media/webrtc/trunk/moz.build b/media/webrtc/trunk/moz.build
--- a/media/webrtc/trunk/moz.build	2018-10-01 13:05:43.719382546 +0100
+++ b/media/webrtc/trunk/moz.build	2018-10-01 17:31:42.168723566 +0100
@@ -252,6 +252,18 @@
         "/media/webrtc/trunk/webrtc/modules/video_processing/video_processing_neon_gn"
     ]
 
+# ensure neon components present for aarch64 Linux
+if CONFIG["CPU_ARCH"] == "aarch64" and CONFIG["OS_TARGET"] == "Linux":
+
+    DIRS += [
+        "/media/webrtc/trunk/webrtc/common_audio/common_audio_neon_c_gn",
+        "/media/webrtc/trunk/webrtc/common_audio/common_audio_neon_gn",
+        "/media/webrtc/trunk/webrtc/modules/audio_coding/isac_neon_gn",
+        "/media/webrtc/trunk/webrtc/modules/audio_processing/audio_processing_neon_c_gn",
+        "/media/webrtc/trunk/webrtc/modules/audio_processing/audio_processing_neon_gn",
+        "/media/webrtc/trunk/webrtc/modules/video_processing/video_processing_neon_gn"
+    ]
+
 if CONFIG["CPU_ARCH"] == "arm" and CONFIG["OS_TARGET"] == "NetBSD":
 
     DIRS += [
diff -uNr a/media/webrtc/trunk/webrtc/common_audio/common_audio_c_gn/moz.build b/media/webrtc/trunk/webrtc/common_audio/common_audio_c_gn/moz.build
--- a/media/webrtc/trunk/webrtc/common_audio/common_audio_c_gn/moz.build	2018-10-01 13:05:43.789381037 +0100
+++ b/media/webrtc/trunk/webrtc/common_audio/common_audio_c_gn/moz.build	2018-10-01 13:30:31.227200556 +0100
@@ -196,11 +196,13 @@
     DEFINES["WEBRTC_ARCH_ARM64"] = True
     DEFINES["WEBRTC_HAS_NEON"] = True
 
-    UNIFIED_SOURCES += [
-        "/media/webrtc/trunk/webrtc/common_audio/signal_processing/complex_bit_reverse.c",
-        "/media/webrtc/trunk/webrtc/common_audio/signal_processing/filter_ar_fast_q12.c",
-        "/media/webrtc/trunk/webrtc/common_audio/signal_processing/spl_sqrt_floor.c"
-    ]
+    # Linux already pulls the following sources in unconditionally
+    if CONFIG["OS_TARGET"] != "Linux":
+        UNIFIED_SOURCES += [
+            "/media/webrtc/trunk/webrtc/common_audio/signal_processing/complex_bit_reverse.c",
+            "/media/webrtc/trunk/webrtc/common_audio/signal_processing/filter_ar_fast_q12.c",
+            "/media/webrtc/trunk/webrtc/common_audio/signal_processing/spl_sqrt_floor.c"
+        ]
 
 if CONFIG["CPU_ARCH"] == "arm":
 
diff -uNr a/media/webrtc/trunk/webrtc/modules/audio_coding/isac_fix_c_gn/moz.build b/media/webrtc/trunk/webrtc/modules/audio_coding/isac_fix_c_gn/moz.build
--- a/media/webrtc/trunk/webrtc/modules/audio_coding/isac_fix_c_gn/moz.build	2018-10-01 13:05:43.779381251 +0100
+++ b/media/webrtc/trunk/webrtc/modules/audio_coding/isac_fix_c_gn/moz.build	2018-10-01 13:30:44.116918435 +0100
@@ -206,13 +206,15 @@
     DEFINES["WEBRTC_ARCH_ARM64"] = True
     DEFINES["WEBRTC_HAS_NEON"] = True
 
-    SOURCES += [
-        "/media/webrtc/trunk/webrtc/modules/audio_coding/codecs/isac/fix/source/pitch_filter_c.c"
-    ]
+    # Linux already pulls the following sources in unconditionally
+    if CONFIG["OS_TARGET"] != "Linux":
+        SOURCES += [
+            "/media/webrtc/trunk/webrtc/modules/audio_coding/codecs/isac/fix/source/pitch_filter_c.c"
+        ]
 
-    UNIFIED_SOURCES += [
-        "/media/webrtc/trunk/webrtc/modules/audio_coding/codecs/isac/fix/source/lattice_c.c"
-    ]
+        UNIFIED_SOURCES += [
+            "/media/webrtc/trunk/webrtc/modules/audio_coding/codecs/isac/fix/source/lattice_c.c"
+        ]
 
 if CONFIG["CPU_ARCH"] == "arm":
 
diff -uNr a/media/webrtc/trunk/webrtc/modules/audio_processing/audio_processing_c_gn/moz.build b/media/webrtc/trunk/webrtc/modules/audio_processing/audio_processing_c_gn/moz.build
--- a/media/webrtc/trunk/webrtc/modules/audio_processing/audio_processing_c_gn/moz.build	2018-10-01 13:05:43.759381684 +0100
+++ b/media/webrtc/trunk/webrtc/modules/audio_processing/audio_processing_c_gn/moz.build	2018-10-01 13:31:07.996395775 +0100
@@ -168,11 +168,13 @@
     DEFINES["WEBRTC_ARCH_ARM64"] = True
     DEFINES["WEBRTC_HAS_NEON"] = True
 
-    UNIFIED_SOURCES += [
-        "/media/webrtc/trunk/webrtc/modules/audio_processing/ns/noise_suppression_x.c",
-        "/media/webrtc/trunk/webrtc/modules/audio_processing/ns/nsx_core.c",
-        "/media/webrtc/trunk/webrtc/modules/audio_processing/ns/nsx_core_c.c"
-    ]
+    # Linux already pulls in a generic set, with differing prototypes
+    if CONFIG["OS_TARGET"] != "Linux":
+        UNIFIED_SOURCES += [
+            "/media/webrtc/trunk/webrtc/modules/audio_processing/ns/noise_suppression_x.c",
+            "/media/webrtc/trunk/webrtc/modules/audio_processing/ns/nsx_core.c",
+            "/media/webrtc/trunk/webrtc/modules/audio_processing/ns/nsx_core_c.c"
+        ]
 
 if CONFIG["CPU_ARCH"] == "arm":
 
